### ISSUE TITLE
API to build reusable FormField could be improved or documented 

### ISSUE THREAD

**Issue by cedvdb** [flutter/flutter#88746]
Forms often contain subset of fields that are one unit. One example of this is an Address field, it contains a street field, box, country, etc. 

I believe the API of FormField makes it hard to reason about such fields. That, or there is another way to do it than the following and it's not documented. 

Here is an example of making an address field (which is not even complete):

<details> 
<summary>minimal code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Home(),
      ),
    );
  }
}

class Home extends StatefulWidget {
  @override
  State<Home> createState() => _HomeState();
}

class _HomeState extends State<Home> {
  late final ValueNotifier<Address?> controller;

  @override
  void initState() {
    super.initState();
    controller = ValueNotifier(null);
    controller.addListener(() => setState(() {}));
  }

  @override
  void dispose() {
    super.dispose();
    controller.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Column(
        children: [
          AddressFormField(controller: controller),
          const SizedBox(height: 8),
          const Text('value:'),
          Text(controller.value.toString()),
          const SizedBox(height: 8),
          ElevatedButton(
              onPressed: () => controller.value = null,
              child: const Text('reset')),
        ],
      ),
    );
  }
}

class Address {
  final String street;
  final String box;
  const Address({
    required this.street,
    required this.box,
  });

  @override
  String toString() => 'Address(street: $street, box: $box)';
}

class AddressFormField extends FormField<Address> {
  @override
  final Address? initialValue;
  final ValueNotifier<Address?>? controller;

  AddressFormField({
    this.initialValue,
    this.controller,
  }) : super(builder: (field) {
          final state = field as AddressFormFieldState;
          return state.builder();
        });

  @override
  AddressFormFieldState createState() => AddressFormFieldState();
}

class AddressFormFieldState extends FormFieldState<Address> {
  late final TextEditingController streetController;
  late final TextEditingController boxController;
  late final ValueNotifier<Address?> controller;

  @override
  AddressFormField get widget => super.widget as AddressFormField;

  @override
  void initState() {
    super.initState();
    streetController = TextEditingController(text: widget.initialValue?.street);
    boxController = TextEditingController(text: widget.initialValue?.box);
    controller =
        widget.controller ?? ValueNotifier<Address?>(widget.initialValue);
    // those two listeners are here so we can reflect the changes from the
    // child inputs (text) to the parent (address).
    streetController.addListener(_reflectChangesOnParent);
    boxController.addListener(_reflectChangesOnParent);

    // this listener is here to reflect the change from the parent to the children
    controller.addListener(_reflectChangesOnChildren);
    // this listener is here to reflect the change on the FormField
    controller.addListener(_onChange);
  }

  @override
  void dispose() {
    super.dispose();
    streetController.dispose();
    boxController.dispose();
    controller.dispose();
  }

  _reflectChangesOnParent() {
    final street = streetController.text;
    final box = boxController.text;
    if (controller.value?.street != street || controller.value?.box != box) {
      controller.value = Address(street: street, box: box);
    }
  }

  _reflectChangesOnChildren() {
    final street = controller.value?.street;
    final box = controller.value?.box;
    if (street != streetController.text) {
      streetController.text = street ?? '';
    }
    if (box != boxController.text) {
      boxController.text = box ?? '';
    }
  }

  _onChange() {
    if (controller.value != value) {
      didChange(controller.value);
    }
  }

  Widget builder() {
    return Column(
      children: [
        TextFormField(
          controller: streetController,
        ),
        TextFormField(
          controller: boxController,
        ),
      ],
    );
  }
}
``` 

</details>

The chit chat between the controllers makes it a bit hard to follow, here there is 2 fields so that's allright, but it quickly becomes annoying.

**Comment by TahaTesser** [comment]
Hi @cedvdb 

> The chit chat between the controllers makes it a bit hard to follow, here there is 2 fields so that's allright, but it quickly becomes annoying.


Can you please provide some examples of this? Please explain in detail

**Comment by cedvdb** [comment]
It's in my example, the methods `_reflectChangesOnParent` and `_reflectChangesOnChildren`.
